torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp

```diff
@@ -0,0 +1,314 @@
+  THDTensor_(copy)(r_, t);
+}
+
+void THDTensor_(sort)(THDTensor *rt_, THDLongTensor *ri_,
+                             THDTensor *t, int dimension,
+                             int descendingOrder) {
+  THArgCheck(dimension >= 0 && dimension < THDTensor_(nDimension)(t),
+      2, "invalid dimension %d", dimension + TH_INDEX_BASE);
+
+  THDTensor_(resizeAs)(rt_, t);
+  THDTensor_(copy)(rt_, t);
+
+  {
+    THLongStorage *size = THDTensor_(newSizeOf)(t);
+    THDLongTensor_resize(ri_, size, NULL);
+    THLongStorage_free(size);
+  }
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorSort, rt_, ri_, t, dimension, descendingOrder),
+    THDState::s_current_worker
+  );
+}
+
+void THDTensor_(topk)(THDTensor *rt_, THDLongTensor *ri_,
+                      THDTensor *t, long k, int dim,
+                      int dir, int sorted) {
+  int numDims = THDTensor_(nDimension)(t);
+  THArgCheck(dim >= 0 && dim < numDims, 3, "dim not in range");
+
+  long sliceSize = THDTensor_(size)(t, dim);
+  THArgCheck(k > 0 && k <= sliceSize, 2, "k not in range for dimension");
+
+  THLongStorage *topKSize = THDTensor_(newSizeOf)(t);
+  THLongStorage_set(topKSize, dim, k);
+  THDTensor_(resize)(rt_, topKSize, NULL);
+  THDLongTensor_resize(ri_, topKSize, NULL);
+  THLongStorage_free(topKSize);
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorTopk, rt_, ri_, t, k, dim, dir, sorted),
+    THDState::s_current_worker
+  );
+}
+
+void THDTensor_(tril)(THDTensor *r_, THDTensor *t, long k) {
+  THArgCheck(THDTensor_(nDimension)(t) == 2, 1, "expected a matrix");
+
+  THDTensor_(resizeAs)(r_, t);
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorTril, r_, t, k),
+    THDState::s_current_worker
+  );
+}
+
+void THDTensor_(triu)(THDTensor *r_, THDTensor *t, long k) {
+  THArgCheck(THDTensor_(nDimension)(t) == 2, 1, "expected a matrix");
+
+  THDTensor_(resizeAs)(r_, t);
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorTriu, r_, t, k),
+    THDState::s_current_worker
+  );
+}
+void THDTensor_(cat)(THDTensor *r_, THDTensor *ta, THDTensor *tb, int dimension) {
+  THDTensor* inputs[2];
+  inputs[0] = ta;
+  inputs[1] = tb;
+  THDTensor_(catArray)(r_, inputs, 2, dimension);
+}
+
+void THDTensor_(catArray)(THDTensor *result, THDTensor **inputs,
+                          int numInputs, int dimension) {
+  THLongStorage *size;
+  long offset;
+  int ndim = dimension + 1;
+  int ldimension = dimension;
+  bool allEmpty = true;
+  for (int i = 0; i < numInputs; i++)
+    ndim = THMax(ndim, inputs[i]->nDimension);
+
+  if (dimension == -2)
+    ldimension = ndim ? (ndim - 1) : 0;
+
+  THArgCheck(numInputs > 0, 3, "invalid number of inputs %d", numInputs);
+  THArgCheck(ldimension >= 0, 4, "invalid dimension %d", dimension + TH_INDEX_BASE);
+
+  size = THLongStorage_newWithSize(ndim);
+
+  for (int i = 0; i < ndim; i++) {
+    long dimSize = i < inputs[0]->nDimension ? inputs[0]->size[i] : 1;
```

Comment:
`: 1` is not correct. You can see that in `TH` there is applied `THMin` at the end: https://github.com/apaszke/pytorch-dist/blob/thd/torch/lib/TH/generic/THTensorMath.c#L2069